### PR TITLE
allow initialization of shared variables

### PR DESCRIPTION
**Pull request by WalterBright**
Partial fix of https://issues.dlang.org/show_bug.cgi?id=20195

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
Please add a bugzilla reference/change log entry

**Review by MoonlightSentinel — APPROVED** [review]

**Review by PetarKirov — APPROVED** [review]

File: changelog/shared-init.dd [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Allow intialization of shared variables.
```
Comment (Geod24) [inline changelog/shared-init.dd:1]:
```suggestion
Allow initialization of shared variables.
```

File: changelog/shared-init.dd [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+Shared variables can now be initialized when the -preview=nosharedaccess is used.
```
Comment (Geod24) [inline changelog/shared-init.dd:3]:
```suggestion
Shared variables can now be initialized when the `-preview=nosharedaccess` switch is used.
```

**Review by Geod24 — APPROVED** [review]
Tested locally, LGTM

**Comment by Geod24** [comment]
@WalterBright : You can use "Issue 20195: allow initialization of shared variables" to reference the issue without closing it.

**Comment by TurkeyMan** [comment]
Does it work with aggregates? Can we get a unit test validating such a struct construction?

**Comment by MoonlightSentinel** [comment]
Just checked, the current patch fails for non-zero initialized aggregates:

```
struct S
{
    int i = 2;
}

interface I
{
    void foo();
}

class C : I
{
    int i;
    override void foo() {}
}

union U
{
    int i = 1;
    bool b;
}

void main()
{
    shared int num;
    shared S s;
    shared I i;
    shared C c;
    shared U u;
}
```
Before:
```
shared_init.d(25): Error: direct access to shared num is not allowed, see core.atomic
shared_init.d(1): Error: direct access to shared S is not allowed, see core.atomic
shared_init.d(27): Error: direct access to shared i is not allowed, see core.atomic
shared_init.d(28): Error: direct access to shared c is not allowed, see core.atomic
shared_init.d(17): Error: direct access to shared U is not allowed, see core.atomic
```

After:
```
shared_init.d(1): Error: direct access to shared S is not allowed, see core.atomic
shared_init.d(17): Error: direct access to shared U is not allowed, see core.atomic
```

**Comment by MoonlightSentinel** [comment]
Ditto, just noticed that dmd didn't issue an error for S if it was zero-initialised.

**Comment by WalterBright** [comment]
Please file bugzilla issues for any new issues.

**Comment by MoonlightSentinel** [comment]
https://issues.dlang.org/show_bug.cgi?id=20908

**Pull request merged**